src/plugins/collapsibleColumns/collapsibleColumns.js

```diff
@@ -408,53 +440,131 @@ class CollapsibleColumns extends BasePlugin {
+        (this.collapsableCoordsList[currentCoords.row] && !this.collapsableCoordsList[currentCoords.row][currentCoords.col])) {
+        collapsePossible = false;
 
-          break;
-        case 'expand':
-          if (this.hiddenColumnsPlugin.isHidden(colToHide)) {
-            hiddenColumns.splice(hiddenColumns.indexOf(colToHide), 1);
-          }
+        return false;
+      }
+
+      collapsePossible = this.collapsableCoordsList[currentCoords.row][currentCoords.col];
+
+      const colspanArray = this.nestedHeadersPlugin.colspanArray;
+      const level = this.nestedHeadersPlugin.rowCoordsToLevel(currentCoords.row);
+      const currentHeaderColspan = colspanArray[level][currentCoords.col].colspan;
+      const childHeaders = this.nestedHeadersPlugin.getChildHeaders(currentCoords.row, currentCoords.col);
+      let nextLevel = level + 1;
+      let childColspanLevel = colspanArray[nextLevel];
+      let firstChildColspan = childColspanLevel ? childColspanLevel[childHeaders[0]].colspan || 1 : 1;
 
-          break;
-        default:
-          break;
+      while (firstChildColspan === currentHeaderColspan && nextLevel < this.columnHeaderLevelCount) {
+        nextLevel += 1;
+        childColspanLevel = colspanArray[nextLevel];
+        firstChildColspan = childColspanLevel ? childColspanLevel[childHeaders[0]].colspan || 1 : 1;
       }
+
+      rangeEach(firstChildColspan, currentHeaderColspan - 1, (i) => {
+        const colToHide = currentCoords.col + i;
+
+        switch (action) {
+          case 'collapse':
+            if (!this.hiddenColumnsPlugin.isHidden(colToHide)) {
+              hiddenColumns.push(colToHide);
+            }
+
+            destinationCollapsedColumns = Array.from(hiddenColumns);
+
+            this.markSectionAs('collapsed', currentCoords.row, currentCoords.col, true);
+
+            break;
+          case 'expand':
+            if (this.hiddenColumnsPlugin.isHidden(colToHide)) {
+              hiddenColumns.splice(hiddenColumns.indexOf(colToHide), 1);
+            }
+
+            destinationCollapsedColumns = Array.from(hiddenColumns);
+
+            this.markSectionAs('expanded', currentCoords.row, currentCoords.col, true);
+
+            break;
+          default:
+            break;
+        }
+      });
     });
 
+    if (action === 'collapse') {
+      const allowColumnCollapse = this.hot.runHooks('beforeColumnCollapse', currentCollapsedColumns, destinationCollapsedColumns, collapsePossible);
+
+      if (allowColumnCollapse === false) {
+        hiddenColumns.length = 0;
+        this.collapsedSections = {};
+
+        return;
+      }
+    } else {
+      const allowColumnExpand = this.hot.runHooks('beforeColumnExpand', currentCollapsedColumns, destinationCollapsedColumns, collapsePossible);
+
+      if (allowColumnExpand === false) {
+        this.hiddenColumnsPlugin.hiddenColumns = Array.from(this.collapsedColumns);
+
+        arrayEach(Object.keys(this.collapsedSections), (key) => {
```

Comment:
The `key` variable name should be more precise about the value it holds. In this case as `rowIndex`?

Reply:
Yes, it is `rowIndex`.